### ISSUE TITLE
Socket hangs up before handshake response when using Node's http server

### ISSUE THREAD

**Issue by prashcr** [uWebSockets/uWebSockets#398]
Include the following information (where applicable):

* What operating system are you running?
  * Windows 10
* What platform/language are you using?
  * Node.js 7.3.0
  * uws ^0.12.0
* Are you using SSL, if so, how?
  * SSL is not part of the equation
* Can you show me some short code that triggers your issue?

**server.js**
```
'use strict';

const WebSocket = require('uws');
const http = require('http');

const wss = new WebSocket.Server({ noServer: true });
const server = http.createServer();

server.on('upgrade', (request, socket, head) => {
  setTimeout(() => {
    wss.handleUpgrade(request, socket, head, () => console.log('connection'));
  }, 100);
});

server.listen(3000, () => console.log('listening on *:3000'));
```

**client-side**
Either open a browser to [http://localhost:3000]() and run the following in the console or use `wscat` to connect.
```
const ws = new WebSocket('ws://localhost:3000')
```  
  
You should get an error about connection closing before handshake response (browser), or socket hanging up (`wscat`)

---
  
I've confirmed that
1. This only occurs on Windows, could not reproduce on OS X
2. When `uws` is replaced with `ws`, it works.
3. Starting a server directly with `uws` also works 
e.g.
```
const wss = new WebSocket.Server({ port: 3000 })
```
The problem seems to be in attaching to Node's http server on Windows.

**Comment by DiegoRBaquero** [comment]
Why not use...
````js
const http = require('http');
const WebSocketServer = require('uws').Server;
const server = http.createServer();
const wss = new WebSocketServer({ server: server });

wss.on('connection', function(ws) {
    ...
});
````

**Comment by prashcr** [comment]
Hmm, that works.

The problem is that I'm facing this while using [Primus](https://github.com/primus/primus), and this seems to be how it handles attaching to Node's http server.
https://github.com/primus/primus/blob/0458cfdeaae0e80df310bb97d0a63e8fddd52c3f/transformers/websockets/server.js#L34

**Comment by alexhultman** [comment]
#338

**Comment by alexhultman** [comment]
If you call handleUpgrade directly without any timeout then I assume it works?

**Comment by prashcr** [comment]
Yep, that's correct

**Comment by alexhultman** [comment]
Does the connection close before handleUpgrade is called? What if you increase the timeout to 5 seconds - does the browser get hung up before the timer fires? You can add text logging to track what happens.

**Comment by prashcr** [comment]
The connection gets closed at the same time as handleUpgrade, regardless of the timeout.

Browser remains responsive, error shows up when handleUpgrade is called.

Here's my modified snippet
```
'use strict';

const WebSocket = require('uws');
const http = require('http');

const wss = new WebSocket.Server({ noServer: true });
const server = http.createServer();

server.on('upgrade', (request, socket, head) => {
  console.log('Upgrade request')
  setTimeout(() => {
    console.log('Handling upgrade')
    wss.handleUpgrade(request, socket, head, () => console.log('connection'));
  }, 5000);
});

server.listen(3000, () => console.log('listening on *:3000'));
```

**Comment by alexhultman** [comment]
If you also print socket._handle.fd right before handleUpgrade that would be nice, last question :wink:

**Comment by prashcr** [comment]
It says `-1`

**Comment by alexhultman** [comment]
Unfortunately this is an issue with nodejs. I have reduced the bug down to this:

```javascript
'use strict';

const uws = require('./dist/uws');
const http = require('http');
const server = http.createServer();

server.on('upgrade', (request, socket, head) => {
  console.log('Got upgrade request');

  // this duplicates the socket
  uws.native.transfer(socket._handle, null);

  // destroying here works fine, leaves duped socket alive
  // socket.destroy();

  setTimeout(() => {
    // destroying here closes the duped socket
    socket.destroy();
  }, 2000);
});

server.listen(3000);
```

Calling socket.destroy directly from the upgrade callback works as it should, it leaves the duped socket as is but calling socket.destroy from within the timer 2 seconds later causes the duped socket to close. They have varying behavior in their destroy function depending on where it is called from which is strange.

I don't know if socket.destroy sends FIN in some cases and thus causing the client to close or if it is something else but the main issue is within nodejs itself. They need to either fix this (which they probably will refuse to do) or provide a documented way of transferring sockets (which they probably will refuse to do).

Nodejs camp is not very easy to work with you know.

**Comment by prashcr** [comment]
Hmm, I found your issues on the topic

https://github.com/nodejs/node/issues/7627
https://github.com/nodejs/node/issues/8974

Anyways, thanks for the help

**Comment by alexhultman** [comment]
Yes exactly. 7627 is already worked around and works but 8974 is what is needed here (to officially support what I do). They don't like what I do, though.

**Comment by alexhultman** [comment]
Ah, splendid, I'm blocked from commenting or filing issues in nodejs camp. Well that's perfect then.

**Comment by alexhultman** [comment]
I did some further digging and this is the flow:

socket.destroy() calls socket._handle.close() which calls uv_close on the uv_tcp_t which calls (at least) close() on the native SOCKET handle. I can trigger the problem by calling uv_close directly, but not by calling close() directly, which is good news because then the problem is somewhere between uv_close and close.

**Comment by prashcr** [comment]
According to @lpinca, your example is slightly off because the 2nd `socket.destroy()` apparently happens on the same tick as `uws.native.transfer()`

**Comment by alexhultman** [comment]
Fixed it. Apparently libuv has its own logic for duping sockets that misbehaves if the uv_tcp_t is closed (uv_close) without having a specific "duplicated socket" flag set. Setting it, `handle->flags |= UV_HANDLE_SHARED_TCP_SOCKET;` fixes the problem.

Should probably swap over to using the (undocumented) uv_tcp_duplicate_socket function instead.

**Comment by alexhultman** [comment]
Fixed in 24ad90c689fef123675f7a8cf9145a6e9f851371 should be in v0.13.0
